File: Tyche.Android/session/src/main/java/com/felipearpa/tyche/session/authentication/infrastructure/AuthenticationRemoteRepository.kt
package com.felipearpa.tyche.session.authentication.infrastructure

import com.felipearpa.network.NetworkExceptionHandler
import com.felipearpa.tyche.session.AccountBundle
import com.felipearpa.tyche.session.authentication.domain.AccountLink
import com.felipearpa.tyche.session.authentication.domain.AuthenticationExternalDataSource
import com.felipearpa.tyche.session.authentication.domain.AuthenticationRemoteDataSource
import com.felipearpa.tyche.session.authentication.domain.AuthenticationRepository
import com.felipearpa.tyche.session.authentication.domain.ExternalAccountId
import com.felipearpa.tyche.session.authentication.domain.SignInWithEmailLinkException
import com.felipearpa.tyche.session.authentication.domain.toLinkAccountRequest
import javax.inject.Inject

internal class AuthenticationRemoteRepository @Inject constructor(
    private val authenticationExternalDataSource: AuthenticationExternalDataSource,
    private val authenticationRemoteDataSource: AuthenticationRemoteDataSource,
    private val networkExceptionHandler: NetworkExceptionHandler,
) : AuthenticationRepository {
    override suspend fun sendSignInLinkToEmail(email: String) =
        handleFirebaseSendSignInLinkToEmail {
            authenticationExternalDataSource.sendSignInLinkToEmail(email = email)
        }

    override suspend fun signInWithEmailLink(
        email: String,
        emailLink: String,
    ): Result<ExternalAccountId> {
        if (!authenticationExternalDataSource.isSignInWithEmailLink(emailLink = emailLink))
            return Result.failure(SignInWithEmailLinkException.InvalidEmailLink)

        return handleFirebaseSignInWithEmailLink {
            authenticationExternalDataSource.signInWithEmailLink(
                email = email,
                emailLink = emailLink,
            )
        }
    }

    override suspend fun signInWithEmailAndPassword(
        email: String,
        password: String,
    ): Result<ExternalAccountId> {
        return handleFirebaseSignInWithEmailAndPassword {
            authenticationExternalDataSource.signInWithEmailAndPassword(
                email = email,
                password = password,
            )
        }
    }

    override suspend fun logout(): Result<Unit> {
        return try {
            authenticationExternalDataSource.signOut()
            Result.success(Unit)
        } catch (exception: Exception) {
            Result.failure(exception)
        }
    }

    override suspend fun linkAccount(accountLink: AccountLink): Result<AccountBundle> =
        networkExceptionHandler.handle {
            authenticationRemoteDataSource.linkAccount(request = accountLink.toLinkAccountRequest())
                .run {
                    AccountBundle(
                        accountId = this.accountId,
                        externalAccountId = accountLink.externalAccountId,
                    )
                }
        }
}

File: Tyche.iOS/Session/Sources/Session/Authentication/Infrastructure/AuthenticationRemoteRepository.swift
import Core

class AuthenticationRemoteRepository: AuthenticationRepository {
    private let authenticationExternalDataSource: AuthenticationExternalDataSource
    private let authenticationRemoteDataSource: AuthenticationRemoteDataSource
    private let networkErrorHandler: NetworkErrorHandler
    
    init(
        authenticationExternalDataSource: AuthenticationExternalDataSource,
        authenticationRemoteDataSource: AuthenticationRemoteDataSource,
        networkErrorHandler: NetworkErrorHandler
    ) {
        self.authenticationExternalDataSource = authenticationExternalDataSource
        self.authenticationRemoteDataSource = authenticationRemoteDataSource
        self.networkErrorHandler = networkErrorHandler
    }
    
    func sendSignInLinkToEmail(email: String) async -> Result<Void, Error> {
        return await handleFirebaseSignInWithEmail {
            try await authenticationExternalDataSource.sendSignInLinkToEmail(email: email)
        }
    }
    
    func signInWithEmailLink(email: String, emailLink: String) async -> Result<ExternalAccountId, Error> {
        if await !authenticationExternalDataSource.isSignInWithEmailLink(emailLink: emailLink) {
            return Result.failure(EmailLinkSignInError.invalidEmailLink)
        }
        
        return await handleFirebaseSignInWithEmailLink {
            try await authenticationExternalDataSource.signInWithEmailLink(email: email, emailLink: emailLink)
        }
    }
    
    func logout() async -> Result<Void, Error> {
        do {
            try await authenticationExternalDataSource.signOut()
            return Result.success(())
        } catch let error {
            return Result.failure(error)
        }
    }
    
    func linkAccount(accountLink: AccountLink) async -> Result<AccountBundle, Error> {
        return await networkErrorHandler.handle {
            let linkAccountResponse = try await authenticationRemoteDataSource.linkAccount(request: accountLink.toLinkAccountRequest())
            return AccountBundle(accountId: linkAccountResponse.accountId, externalAccountId: accountLink.externalAccountId)
        }
    }
}
